pkg/querier/querysharding/queryable.go

```diff
@@ -0,0 +1,105 @@
+package querysharding
+
+import (
+	"context"
+	"encoding/hex"
+	"github.com/cortexproject/cortex/pkg/querier/astmapper"
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+const (
+	missingEmbeddedQueryMsg = "missing embedded query"
+	nonEmbeddedErrMsg       = "DownstreamQuerier cannot handle a non-embedded query"
+)
+
+// DownstreamQueryable is a wrapper for and implementor of the Queryable interface.
+type DownstreamQueryable struct {
+	Req     *queryrange.Request
+	Handler queryrange.Handler
+}
+
+func (q *DownstreamQueryable) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+	return &DownstreamQuerier{ctx, q.Req, q.Handler}, nil
+}
+
+// DownstreamQueryable is a an implementor of the Queryable interface.
+type DownstreamQuerier struct {
+	Ctx     context.Context
+	Req     *queryrange.Request
+	Handler queryrange.Handler
+}
+
+// Select returns a set of series that matches the given label matchers.
+func (q *DownstreamQuerier) Select(
+	_ *storage.SelectParams,
+	matchers ...*labels.Matcher,
+) (storage.SeriesSet, storage.Warnings, error) {
+	var embeddedQuery string
+	var isEmbedded bool
+	for _, matcher := range matchers {
+		if matcher.Name == "__name__" && matcher.Value == astmapper.EMBEDDED_QUERY_FLAG {
+			isEmbedded = true
+		}
+
+		if matcher.Name == astmapper.QUERY_LABEL {
+			embeddedQuery = matcher.Value
+		}
+	}
+
+	if isEmbedded {
+		if embeddedQuery != "" {
+			return q.handleEmbeddedQuery(embeddedQuery)
+		} else {
+			return nil, nil, errors.Errorf(missingEmbeddedQueryMsg)
+		}
+
+	}
+
+	return nil, nil, errors.Errorf(nonEmbeddedErrMsg)
+}
+
+// handleEmbeddedQuery defers execution of an encoded query to a downstream Handler
+func (q *DownstreamQuerier) handleEmbeddedQuery(encoded string) (storage.SeriesSet, storage.Warnings, error) {
+	decoded, err := hex.DecodeString(encoded)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resp, err := q.Handler.Do(q.Ctx, ReplaceQuery(*q.Req, string(decoded)))
```

Comment:
This is going to be called sequentially by the engine. You need to implement a lazy `storage.SeriesSet` that doesn't block the engine waiting for the result.

Reply:
Thanks for the heads up. This should be resolved by 46f7dfe and 9c15c0f!